File: src/a1/mood_detector.py
"""Mood detection module for A1.

Analyses text (and optionally numeric valence/energy scores) to return
a normalised :class:`Mood` object that drives lyrics and music generation.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum

# ---------------------------------------------------------------------------
# Mood taxonomy
# ---------------------------------------------------------------------------


class MoodCategory(str, Enum):
    """High-level emotional categories recognised by A1."""

    HAPPY = "happy"
    SAD = "sad"
    REFLECTIVE = "reflective"
    ANXIOUS = "anxious"
    ENERGIZED = "energized"
    CALM = "calm"
    ROMANTIC = "romantic"
    INSPIRED = "inspired"


@dataclass
class Mood:
    """A detected or user-supplied emotional state.

    Attributes:
        category:   The primary emotional category.
        intensity:  Strength of the emotion on a 0.0–1.0 scale.
        description: Optional free-text description supplied by the user.
        keywords:   Words that contributed to the detection result.
    """

    category: MoodCategory
    intensity: float = 0.7
    description: str | None = None
    keywords: list[str] = field(default_factory=list)

    def __post_init__(self) -> None:
        if not 0.0 <= self.intensity <= 1.0:
            raise ValueError("intensity must be between 0.0 and 1.0")

    @property
    def is_upbeat(self) -> bool:
        """Return True for moods that call for energetic/bright music."""
        return self.category in {
            MoodCategory.HAPPY,
            MoodCategory.ENERGIZED,
            MoodCategory.INSPIRED,
        }

    @property
    def is_introspective(self) -> bool:
        """Return True for moods that call for gentle, thoughtful music."""
        return self.category in {
            MoodCategory.REFLECTIVE,
            MoodCategory.SAD,
            MoodCategory.CALM,
            MoodCategory.ANXIOUS,
        }


# ---------------------------------------------------------------------------
# Keyword bank used for text-based detection
# ---------------------------------------------------------------------------

_KEYWORD_MAP: dict[MoodCategory, list[str]] = {
    MoodCategory.HAPPY: [
        "happy",
        "joy",
        "joyful",
        "excited",
        "elated",
        "cheerful",
        "great",
        "wonderful",
        "amazing",
        "fantastic",
        "laugh",
        "smile",
        "fun",
        "celebrate",
        "glad",
        "delighted",
        "pleased",
    ],
    MoodCategory.SAD: [
        "sad",
        "cry",
        "crying",
        "tears",
        "depressed",
        "heartbroken",
        "unhappy",
        "miserable",
        "grief",
        "lonely",
        "lost",
        "down",
        "hopeless",
        "hurt",
        "pain",
        "sorrow",
        "sorrowful",
    ],
    MoodCategory.REFLECTIVE: [
        "reflect",
        "thinking",
        "nostalgic",
        "memories",
        "remember",
        "miss",
        "wonder",
        "ponder",
        "contemplat",
        "peaceful",
        "quiet",
        "thoughtful",
    ],
    MoodCategory.ANXIOUS: [
        "anxious",
        "anxiety",
        "nervous",
        "stress",
        "stressed",
        "worried",
        "worry",
        "fear",
        "scared",
        "overwhelm",
        "panic",
        "restless",
        "uneasy",
    ],
    MoodCategory.ENERGIZED: [
        "energized",
        "pumped",
        "motivated",
        "fired up",
        "hyped",
        "ready",
        "power",
        "strong",
        "unstoppable",
        "determined",
        "driven",
        "alive",
    ],
    MoodCategory.CALM: [
        "calm",
        "relax",
        "relaxed",
        "peaceful",
        "serene",
        "tranquil",
        "chill",
        "mellow",
        "ease",
        "gentle",
        "soothed",
        "still",
    ],
    MoodCategory.ROMANTIC: [
        "love",
        "romantic",
        "romance",
        "heart",
        "affection",
        "crush",
        "darling",
        "beloved",
        "together",
        "close",
        "intimate",
        "adore",
    ],
    MoodCategory.INSPIRED: [
        "inspired",
        "inspire",
        "creative",
        "dream",
        "vision",
        "purpose",
        "hope",
        "believe",
        "aspire",
        "ambition",
        "passion",
        "future",
    ],
}


# ---------------------------------------------------------------------------
# Public detector
# ---------------------------------------------------------------------------


class MoodDetector:
    """Detect mood from user-supplied text using keyword matching.

    For production use this class can be subclassed or replaced with a model
    that calls a sentiment / emotion-analysis API (e.g. Google NL API,
    AWS Comprehend, or a fine-tuned transformer).
    """

    def detect(self, text: str) -> Mood:
        """Return a :class:`Mood` inferred from *text*.

        The method scores each :class:`MoodCategory` by how many of its
        keywords appear in the lower-cased text, then picks the winner.
        A tie is broken in favour of the first category defined in the enum.

        Args:
            text: Free-form text from the user describing how they feel.

        Returns:
            A :class:`Mood` instance populated with the detected category,
            an intensity proportional to the number of keyword hits, and the
            matched keywords.

        Raises:
            ValueError: If *text* is empty or contains only whitespace.
        """
        if not text or not text.strip():
            raise ValueError("text must not be empty")

        text_lower = text.lower()
        scores: dict[MoodCategory, list[str]] = {cat: [] for cat in MoodCategory}

        for category, keywords in _KEYWORD_MAP.items():
            for kw in keywords:
                # Use word-boundary matching so "sad" does not match "sadly"
                # when "sad" itself is the keyword, but sub-string keywords
                # like "contemplat" still fire on "contemplating".
                pattern = r"\b" + re.escape(kw) if " " not in kw else re.escape(kw)
                if re.search(pattern, text_lower):
                    scores[category].append(kw)

        best_category = max(scores, key=lambda c: len(scores[c]))
        matched = scores[best_category]

        # Intensity: scale the hit-count against the keyword bank size, capped
        # at 0.95 so there is always room for a human-supplied boost.
        bank_size = len(_KEYWORD_MAP[best_category])
        raw_intensity = min(len(matched) / max(bank_size, 1), 0.95)
        # Give at least 0.5 even on a single keyword match.
        intensity = max(raw_intensity, 0.5) if matched else 0.5

        return Mood(
            category=best_category,
            intensity=round(intensity, 2),
            description=text.strip(),
            keywords=matched,
        )

    def from_vibe_choice(self, choice: str) -> Mood:
        """Map a user's short vibe label (e.g. "hopeful") to a :class:`Mood`.

        This is used when A1 asks *"What's your vibe today?"* and the user
        picks from a short list rather than typing free text.

        Args:
            choice: A vibe label string (case-insensitive).

        Returns:
            A :class:`Mood` with a default intensity of 0.7.

        Raises:
            ValueError: If *choice* does not map to any known vibe.
        """
        vibe_map: dict[str, MoodCategory] = {
            "hopeful": MoodCategory.INSPIRED,
            "fun": MoodCategory.HAPPY,
            "soothing": MoodCategory.CALM,
            "happy": MoodCategory.HAPPY,
            "sad": MoodCategory.SAD,
            "reflective": MoodCategory.REFLECTIVE,
            "anxious": MoodCategory.ANXIOUS,
            "energized": MoodCategory.ENERGIZED,
            "calm": MoodCategory.CALM,
            "romantic": MoodCategory.ROMANTIC,
            "inspired": MoodCategory.INSPIRED,
        }
        key = choice.strip().lower()
        if key not in vibe_map:
            available = ", ".join(sorted(vibe_map))
            raise ValueError(
                f"Unknown vibe '{choice}'. Available vibes: {available}"
            )
        return Mood(category=vibe_map[key], intensity=0.7, description=choice)

File: src/moodDetector.js
'use strict';

/**
 * A1 Mood Detector
 *
 * Analyses free-form text and returns a scored map of detected moods.
 * Uses a keyword-based scoring approach so no external API is required.
 */

/** @typedef {'happy'|'sad'|'calm'|'energetic'|'angry'|'hopeful'} Mood */

/**
 * Keyword banks for each mood.
 * Each word carries a weight (default 1). Higher weight = stronger signal.
 * @type {Record<string, Array<{word:string, weight:number}>>}
 */
const MOOD_KEYWORDS = {
  happy: [
    { word: 'happy', weight: 2 }, { word: 'joy', weight: 2 }, { word: 'joyful', weight: 2 },
    { word: 'excited', weight: 1.5 }, { word: 'great', weight: 1 }, { word: 'amazing', weight: 1.5 },
    { word: 'wonderful', weight: 1.5 }, { word: 'fantastic', weight: 1.5 }, { word: 'good', weight: 1 },
    { word: 'love', weight: 1 }, { word: 'laugh', weight: 1 }, { word: 'smile', weight: 1 },
    { word: 'celebrate', weight: 1.5 }, { word: 'fun', weight: 1 }, { word: 'cheerful', weight: 2 },
    { word: 'elated', weight: 2 }, { word: 'delighted', weight: 2 }, { word: 'bliss', weight: 2 },
    { word: 'thrilled', weight: 1.5 }, { word: 'upbeat', weight: 1.5 },
  ],
  sad: [
    { word: 'sad', weight: 2 }, { word: 'cry', weight: 2 }, { word: 'crying', weight: 2 },
    { word: 'depressed', weight: 2 }, { word: 'lonely', weight: 2 }, { word: 'miss', weight: 1 },
    { word: 'missing', weight: 1 }, { word: 'heartbroken', weight: 2 }, { word: 'grief', weight: 2 },
    { word: 'grieve', weight: 2 }, { word: 'lost', weight: 1 }, { word: 'hurt', weight: 1.5 },
    { word: 'pain', weight: 1.5 }, { word: 'unhappy', weight: 2 }, { word: 'miserable', weight: 2 },
    { word: 'down', weight: 1 }, { word: 'blue', weight: 1 }, { word: 'tears', weight: 1.5 },
    { word: 'hopeless', weight: 1.5 }, { word: 'broken', weight: 1.5 },
  ],
  calm: [
    { word: 'calm', weight: 2 }, { word: 'relaxed', weight: 2 }, { word: 'peaceful', weight: 2 },
    { word: 'serene', weight: 2 }, { word: 'chill', weight: 1.5 }, { word: 'quiet', weight: 1 },
    { word: 'still', weight: 1 }, { word: 'tranquil', weight: 2 }, { word: 'mellow', weight: 1.5 },
    { word: 'easy', weight: 1 }, { word: 'rest', weight: 1 }, { word: 'breathe', weight: 1 },
    { word: 'meditate', weight: 1.5 }, { word: 'zen', weight: 1.5 }, { word: 'soothing', weight: 1.5 },
  ],
  energetic: [
    { word: 'energetic', weight: 2 }, { word: 'energy', weight: 1.5 }, { word: 'pumped', weight: 2 },
    { word: 'motivated', weight: 2 }, { word: 'workout', weight: 1.5 }, { word: 'exercise', weight: 1.5 },
    { word: 'run', weight: 1 }, { word: 'running', weight: 1 }, { word: 'active', weight: 1 },
    { word: 'power', weight: 1 }, { word: 'strong', weight: 1.5 }, { word: 'focused', weight: 1.5 },
    { word: 'hustle', weight: 1.5 }, { word: 'driven', weight: 1.5 }, { word: 'fired up', weight: 2 },
    { word: 'unstoppable', weight: 2 }, { word: 'hype', weight: 1.5 }, { word: 'hyped', weight: 1.5 },
  ],
  angry: [
    { word: 'angry', weight: 2 }, { word: 'anger', weight: 2 }, { word: 'furious', weight: 2 },
    { word: 'frustrated', weight: 2 }, { word: 'mad', weight: 1.5 }, { word: 'rage', weight: 2 },
    { word: 'hate', weight: 1.5 }, { word: 'annoyed', weight: 1.5 }, { word: 'irritated', weight: 1.5 },
    { word: 'upset', weight: 1 }, { word: 'fed up', weight: 1.5 }, { word: 'livid', weight: 2 },
    { word: 'outraged', weight: 2 }, { word: 'resentful', weight: 1.5 },
  ],
  hopeful: [
    { word: 'hopeful', weight: 2 }, { word: 'hope', weight: 1.5 }, { word: 'optimistic', weight: 2 },
    { word: 'better', weight: 1 }, { word: 'improve', weight: 1 }, { word: 'future', weight: 1 },
    { word: 'believe', weight: 1.5 }, { word: 'faith', weight: 1.5 }, { word: 'grateful', weight: 1.5 },
    { word: 'thankful', weight: 1.5 }, { word: 'inspired', weight: 2 }, { word: 'courage', weight: 1.5 },
    { word: 'brave', weight: 1.5 }, { word: 'determined', weight: 1.5 }, { word: 'persevere', weight: 1.5 },
    { word: 'rise', weight: 1 }, { word: 'overcome', weight: 1.5 }, { word: 'grow', weight: 1 },
  ],
};

/**
 * Detect and score moods present in the provided text.
 *
 * @param {string} text - Free-form text describing how the user feels.
 * @returns {{ scores: Record<string, number>, dominantMood: string, confidence: number }}
 *   `scores`       — normalised score (0–1) per mood.
 *   `dominantMood` — highest-scoring mood.
 *   `confidence`   — normalised score of the dominant mood (0–1).
 */
function detectMood(text) {
  if (typeof text !== 'string' || text.trim() === '') {
    return { scores: {}, dominantMood: 'calm', confidence: 0 };
  }

  const lower = text.toLowerCase();
  const rawScores = {};

  for (const [mood, keywords] of Object.entries(MOOD_KEYWORDS)) {
    let score = 0;
    for (const { word, weight } of keywords) {
      // Use word-boundary matching to avoid partial hits (e.g. "go" inside "good")
      const pattern = new RegExp(`\\b${escapeRegExp(word)}\\b`, 'gi');
      const matches = lower.match(pattern);
      if (matches) {
        score += matches.length * weight;
      }
    }
    rawScores[mood] = score;
  }

  // Normalise to 0-1 range
  const maxScore = Math.max(...Object.values(rawScores), 1);
  const scores = {};
  let dominantMood = 'calm';
  let highestScore = 0;

  for (const [mood, score] of Object.entries(rawScores)) {
    scores[mood] = score / maxScore;
    if (scores[mood] > highestScore) {
      highestScore = scores[mood];
      dominantMood = mood;
    }
  }

  return { scores, dominantMood, confidence: highestScore };
}

/**
 * Escape special regex characters in a string.
 * @param {string} str
 * @returns {string}
 */
function escapeRegExp(str) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

module.exports = { detectMood, MOOD_KEYWORDS };
